Windows + MinGWの環境で、リンク時に_lock_fileと_unlock_fileのシンボルが解決できない

# 状況
GitHub ホステッドランナー(Windows)上でビルドを行うと、
リンク時に_lock_fileと_unlock_fileのシンボルが解決されず、異常終了する。

# 関連
https://github.com/rust-lang/rust/issues/112368 と同根かも。

# エラーメッセージ

```

error: linking with `x86_64-w64-mingw32-gcc` failed: exit code: 1
  |
  = note: "x86_64-w64-mingw32-gcc" "-fno-use-linker-plugin" "-Wl,--dynamicbase" "-Wl,--disable-auto-image-base" "-m64" "-Wl,--high-entropy-va" "C:\\Users\\runneradmin\\.rustup\\toolchains\\stable-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\rsbegin.o" "C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\rustcCiDpta\\symbols.o" "D:\\a\\recisdb-rs\\recisdb-rs\\target\\debug\\deps\\b25_sys-f7261e01c4b20652.16t4eyqcob8a9hat.rcgu.o" "D:\\a\\recisdb-rs\\recisdb-rs\\target\\debug\\deps\\b25_sys-f7261e01c4b20652.1c5cky5cso7xvi0n.rcgu.o" "D:\\a\\recisdb-rs\\recisdb-rs\\target\\debug\\deps\\b25_sys-f7261e01c4b20652.1gq58g936w3ui0cu.rcgu.o" "D:\\a\\recisdb-rs\\recisdb-rs\\target\\debug\\deps\\b25_sys-f7261e01c4b20652.1gwhpx0homgh1i9s.rcgu.o" "D:\\a\\recisdb-rs\\recisdb-rs\\target\\debug\\deps\\b25_sys-f7261e01c4b20652.1v2v4x6hiiknmc22.rcgu.o" "D:\\a\\recisdb-rs\\recisdb-rs\\target\\debug\\deps\\b25_sys-f7261e01c4b20652.1zp3whreinu4qbd9.rcgu.o" "D:\\a\\recisdb-rs\\recisdb-rs\\target\\debug\\deps\\b25_sys-f7261e01c4b20652.24ou48v2uu8t23kz.rcgu.o" "D:\\a\\recisdb-rs\\recisdb-rs\\target\\debug\\deps\\b25_sys-f7261e01c4b20652.2a1a7h86hsqksmpz.rcgu.o" "D:\\a\\recisdb-rs\\recisdb-rs\\target\\debug\\deps\\b25_sys-f7261e01c4b20652.2gkjqyu19p6f4sxw.rcgu.o" "D:\\a\\recisdb-rs\\recisdb-rs\\target\\debug\\deps\\b25_sys-f7261e01c4b20652.2tr8lno714zzn1i9.rcgu.o" "D:\\a\\recisdb-rs\\recisdb-rs\\target\\debug\\deps\\b25_sys-f7261e01c4b20652.2txxkx6iluvgve69.rcgu.o" "D:\\a\\recisdb-rs\\recisdb-rs\\target\\debug\\deps\\b25_sys-f7261e01c4b20652.33g343egzzq52o7a.rcgu.o" "D:\\a\\recisdb-rs\\recisdb-rs\\target\\debug\\deps\\b25_sys-f7261e01c4b20652.3galkroco80ms0x1.rcgu.o" "D:\\a\\recisdb-rs\\recisdb-rs\\target\\debug\\deps\\b25_sys-f7261e01c4b20652.3sa69ixdcl71itha.rcgu.o" "D:\\a\\recisdb-rs\\recisdb-rs\\target\\debug\\deps\\b25_sys-f7261e01c4b20652.46pgoa6wwjfyq0lg.rcgu.o" "D:\\a\\recisdb-rs\\recisdb-rs\\target\\debug\\deps\\b25_sys-f7261e01c4b20652.4bpsq82jht8om6ki.rcgu.o" "D:\\a\\recisdb-rs\\recisdb-rs\\target\\debug\\deps\\b25_sys-f7261e01c4b20652.56zn1eg8a7bif33r.rcgu.o" "D:\\a\\recisdb-rs\\recisdb-rs\\target\\debug\\deps\\b25_sys-f7261e01c4b20652.g5aqhsiunsfdifg.rcgu.o" "D:\\a\\recisdb-rs\\recisdb-rs\\target\\debug\\deps\\b25_sys-f7261e01c4b20652.y5rbh3o4kho3pwr.rcgu.o" "D:\\a\\recisdb-rs\\recisdb-rs\\target\\debug\\deps\\b25_sys-f7261e01c4b20652.1kx5i3ecov0kwnar.rcgu.o" "-L" "D:\\a\\recisdb-rs\\recisdb-rs\\target\\debug\\deps" "-L" "C:\\Windows\\System32" "-L" "D:\\a\\recisdb-rs\\recisdb-rs\\target\\debug\\build\\b25-sys-ae2aa1ffc12fbda0\\out/lib" "-L" "C:\\Users\\runneradmin\\.rustup\\toolchains\\stable-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib" "-Wl,-Bdynamic" "-lucrt" "-lwinscard" "-Wl,-Bstatic" "-Wl,--whole-archive" "-laribb25" "-Wl,--no-whole-archive" "C:\\Users\\runneradmin\\.rustup\\toolchains\\stable-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\libtest-eb65851c522d788e.rlib" "C:\\Users\\runneradmin\\.rustup\\toolchains\\stable-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\libgetopts-61a280a25b41e3b2.rlib" "C:\\Users\\runneradmin\\.rustup\\toolchains\\stable-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\libunicode_width-c85c92b896a9c2f9.rlib" "C:\\Users\\runneradmin\\.rustup\\toolchains\\stable-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\librustc_std_workspace_std-1860a805430f2e31.rlib" "D:\\a\\recisdb-rs\\recisdb-rs\\target\\debug\\deps\\libpin_project_lite-c6a4aabae59852a2.rlib" "D:\\a\\recisdb-rs\\recisdb-rs\\target\\debug\\deps\\liblog-c801d0c7eeedae9f.rlib" "C:\\Users\\runneradmin\\.rustup\\toolchains\\stable-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\libstd-5b7bb87ce5571884.rlib" "C:\\Users\\runneradmin\\.rustup\\toolchains\\stable-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\libpanic_unwind-833364761c2781ff.rlib" "C:\\Users\\runneradmin\\.rustup\\toolchains\\stable-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\libobject-31ed18134a3c5ef8.rlib" "C:\\Users\\runneradmin\\.rustup\\toolchains\\stable-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\libmemchr-e6b6e6948f8fd702.rlib" "C:\\Users\\runneradmin\\.rustup\\toolchains\\stable-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\libaddr2line-8f2f336806363ae9.rlib" "C:\\Users\\runneradmin\\.rustup\\toolchains\\stable-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\libgimli-e293aed9ea892a13.rlib" "C:\\Users\\runneradmin\\.rustup\\toolchains\\stable-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\librustc_demangle-2b64d871f12392af.rlib" "C:\\Users\\runneradmin\\.rustup\\toolchains\\stable-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\libstd_detect-96a374841ff446f9.rlib" "C:\\Users\\runneradmin\\.rustup\\toolchains\\stable-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\libhashbrown-e044a2cf0118811c.rlib" "C:\\Users\\runneradmin\\.rustup\\toolchains\\stable-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\librustc_std_workspace_alloc-30e5ce84f317655b.rlib" "C:\\Users\\runneradmin\\.rustup\\toolchains\\stable-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\libminiz_oxide-7b7803bde19e4862.rlib" "C:\\Users\\runneradmin\\.rustup\\toolchains\\stable-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\libadler-ad9ad6e6c123302e.rlib" "C:\\Users\\runneradmin\\.rustup\\toolchains\\stable-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\libunwind-d6d690b4e8bbc285.rlib" "C:\\Users\\runneradmin\\.rustup\\toolchains\\stable-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\libcfg_if-3efb1be5ec3dfbdc.rlib" "C:\\Users\\runneradmin\\.rustup\\toolchains\\stable-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\liblibc-49ecf7da6013ea00.rlib" "C:\\Users\\runneradmin\\.rustup\\toolchains\\stable-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\liballoc-111544d92a21aec0.rlib" "C:\\Users\\runneradmin\\.rustup\\toolchains\\stable-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\librustc_std_workspace_core-fde3932818a6e910.rlib" "C:\\Users\\runneradmin\\.rustup\\toolchains\\stable-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\libcore-5b50c1fb747ce736.rlib" "C:\\Users\\runneradmin\\.rustup\\toolchains\\stable-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\libcompiler_builtins-3fc553da9dd474fb.rlib" "-Wl,-Bdynamic" "-lkernel32" "-lkernel32" "-ladvapi32" "-lbcrypt" "-lkernel32" "-lntdll" "-luserenv" "-lws2_32" "-lkernel32" "-lgcc_eh" "-l:libpthread.a" "-lmsvcrt" "-lmingwex" "-lmingw32" "-lgcc" "-lmsvcrt" "-luser32" "-lkernel32" "-Wl,--nxcompat" "-L" "C:\\Users\\runneradmin\\.rustup\\toolchains\\stable-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib" "-o" "D:\\a\\recisdb-rs\\recisdb-rs\\target\\debug\\deps\\b25_sys-f7261e01c4b20652.exe" "-Wl,--gc-sections" "-no-pie" "-nodefaultlibs" "C:\\Users\\runneradmin\\.rustup\\toolchains\\stable-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\rsend.o"
  = note: Warning: corrupt .drectve at end of def file

          Warning: corrupt .drectve at end of def file

          Warning: .drectve `-exclude-symbols:"_ZN7b25_sys8bindings12arib_std_b2528bindgen_test_layout___fsid_t28_$u7b$$u7b$closure$u7d$$u7d$17h17e99638fb25aa59E" ' unrecognized

          Warning: .drectve `-exclude-symbols:"_ZN7b25_sys8bindings12arib_std_b2537bindgen_test_layout_B_CAS_INIT_STATUS28_$u7b$$u7b$closure$u7d$$u7d$17h558990ac61e5f19dE" ' unrecognized

          Warning: .drectve `-exclude-symbols:"_ZN7b25_sys8bindings12arib_std_b2528bindgen_test_layout_B_CAS_ID28_$u7b$$u7b$closure$u7d$$u7d$17h40b2c4a2addac600E" ' unrecognized

          Warning: .drectve `-exclude-symbols:"_ZN7b25_sys8bindings12arib_std_b2537bindgen_test_layout_B_CAS_PWR_ON_CTRL28_$u7b$$u7b$closure$u7d$$u7d$17h11c91160d770dec9E" ' unrecognized

          Warning: .drectve `-exclude-symbols:"_ZN7b25_sys8bindings12arib_std_b2542bindgen_test_layout_B_CAS_PWR_ON_CTRL_INFO28_$u7b$$u7b$closure$u7d$$u7d$17ha325fa6d2e598a37E" ' unrecognized

          Warning: .drectve `-exclude-symbols:"_ZN7b25_sys8bindings12arib_std_b2536bindgen_test_layout_B_CAS_ECM_RESULT28_$u7b$$u7b$closure$u7d$$u7d$17ha984cecc23c0f94bE" ' unrecognized

          Warning: .drectve `-exclude-symbols:"_ZN7b25_sys8bindings12arib_std_b2530bindgen_test_layout_B_CAS_CARD28_$u7b$$u7b$closure$u7d$$u7d$17h92672ba7326a16d2E" ' unrecognized

          Warning: .drectve `-exclude-symbols:"_ZN7b25_sys8bindings12arib_std_b2539bindgen_test_layout_ARIB_STD_B25_BUFFER28_$u7b$$u7b$closure$u7d$$u7d$17hcc75c9cee9cdf502E" ' unrecognized

          Warning: .drectve `-exclude-symbols:"_ZN7b25_sys8bindings12arib_std_b2545bindgen_test_layout_ARIB_STD_B25_PROGRAM_INFO28_$u7b$$u7b$closure$u7d$$u7d$17hbd79388248c8cd71E" ' unrecognized

          Warning: .drectve `-exclude-symbols:"_ZN7b25_sys8bindings12arib_std_b2532bindgen_test_layout_ARIB_STD_B2528_$u7b$$u7b$closure$u7d$$u7d$17h70576fd92fff2128E" ' unrecognized

          Warning: corrupt .drectve at end of def file

          Warning: corrupt .drectve at end of def file

          Warning: .drectve `-exclude-symbols:"_ZN5alloc7raw_vec19RawVec$LT$T$C$A$GT$11allocate_in17h34c14d0c669cd358E" ' unrecognized

          Warning: corrupt .drectve at end of def file

          Warning: corrupt .drectve at end of def file

          Warning: corrupt .drectve at end of def file

          Warning: .drectve `-exclude-symbols:_ZN3std2rt10lang_start17h37550e29cd25abd6E ' unrecognized

          Warning: corrupt .drectve at end of def file

          Warning: corrupt .drectve at end of def file

          Warning: .drectve `-exclude-symbols:_ZN7b25_sys8bindings12arib_std_b2528bindgen_test_layout___fsid_t17h360da14c1822e7ebE ' unrecognized

          Warning: .drectve `-exclude-symbols:_ZN7b25_sys8bindings12arib_std_b2537bindgen_test_layout_B_CAS_INIT_STATUS17h4aed1aac10dc504fE ' unrecognized

          Warning: .drectve `-exclude-symbols:_ZN7b25_sys8bindings12arib_std_b2528bindgen_test_layout_B_CAS_ID17hc7fb6de0ab06bfb8E ' unrecognized

          Warning: .drectve `-exclude-symbols:_ZN7b25_sys8bindings12arib_std_b2537bindgen_test_layout_B_CAS_PWR_ON_CTRL17hf56526344368a9bfE ' unrecognized

          Warning: .drectve `-exclude-symbols:_ZN7b25_sys8bindings12arib_std_b2542bindgen_test_layout_B_CAS_PWR_ON_CTRL_INFO17hacd8d9ed5d6355efE ' unrecognized

          Warning: .drectve `-exclude-symbols:_ZN7b25_sys8bindings12arib_std_b2536bindgen_test_layout_B_CAS_ECM_RESULT17ha846894ec1c6fc31E ' unrecognized

          Warning: .drectve `-exclude-symbols:_ZN7b25_sys8bindings12arib_std_b2530bindgen_test_layout_B_CAS_CARD17h216720d10973fcc6E ' unrecognized

          Warning: .drectve `-exclude-symbols:_ZN7b25_sys8bindings12arib_std_b2539bindgen_test_layout_ARIB_STD_B25_BUFFER17hae73fc57a51e1b24E ' unrecognized

          Warning: .drectve `-exclude-symbols:_ZN7b25_sys8bindings12arib_std_b2545bindgen_test_layout_ARIB_STD_B25_PROGRAM_INFO17h463905bcf08dda03E ' unrecognized

          Warning: .drectve `-exclude-symbols:_ZN7b25_sys8bindings12arib_std_b2532bindgen_test_layout_ARIB_STD_B2517hc81f156562c0c11eE ' unrecognized

          Warning: corrupt .drectve at end of def file

          Warning: .drectve `-exclude-symbols:"_ZN4core3ops8function6FnOnce40call_once$u7b$$u7b$vtable.shim$u7d$$u7d$17h2654e4199e2d7756E" ' unrecognized

          Warning: .drectve `-exclude-symbols:_ZN4core3ops8function6FnOnce9call_once17h3b3378a6066b4302E ' unrecognized

          Warning: .drectve `-exclude-symbols:_ZN4core3ops8function6FnOnce9call_once17h46b4a271df8ac4bcE ' unrecognized

          Warning: .drectve `-exclude-symbols:_ZN4core3ops8function6FnOnce9call_once17h4af0713da1e72487E ' unrecognized

          Warning: .drectve `-exclude-symbols:_ZN4core3ops8function6FnOnce9call_once17h51ea60dc1e0267f5E ' unrecognized

          Warning: .drectve `-exclude-symbols:_ZN4core3ops8function6FnOnce9call_once17h6a411a5ef7160e96E ' unrecognized

          Warning: .drectve `-exclude-symbols:_ZN4core3ops8function6FnOnce9call_once17h6ea265f7ae438d45E ' unrecognized

          Warning: .drectve `-exclude-symbols:_ZN4core3ops8function6FnOnce9call_once17h70066ef2a2a9236eE ' unrecognized

          Warning: .drectve `-exclude-symbols:_ZN4core3ops8function6FnOnce9call_once17h752e5015f1e5e62aE ' unrecognized

          Warning: .drectve `-exclude-symbols:_ZN4core3ops8function6FnOnce9call_once17h7cafe00743a572b5E ' unrecognized

          Warning: .drectve `-exclude-symbols:_ZN4core3ops8function6FnOnce9call_once17h83c187478877c283E ' unrecognized

          Warning: .drectve `-exclude-symbols:_ZN4core3ops8function6FnOnce9call_once17ha345efbd415485a7E ' unrecognized

          Warning: .drectve `-exclude-symbols:_ZN4core3ops8function6FnOnce9call_once17hf9fbd77a47b32b64E ' unrecognized

          Warning: .drectve `-exclude-symbols:_ZN4core3ops8function6FnOnce9call_once17hfb04338fd89c682bE ' unrecognized

          Warning: .drectve `-exclude-symbols:"_ZN4core3ptr30drop_in_place$LT$$RF$usize$GT$17h46e5dc189c605950E" ' unrecognized

          Warning: .drectve `-exclude-symbols:"_ZN4core3ptr78drop_in_place$LT$std..sync..mutex..Mutex$LT$alloc..vec..Vec$LT$u64$GT$$GT$$GT$17h98b2d6b4b3c49f13E" ' unrecognized

          Warning: corrupt .drectve at end of def file

          Warning: corrupt .drectve at end of def file

          Warning: corrupt .drectve at end of def file

          Warning: .drectve `-exclude-symbols:_ZN4core3fmt9Arguments6as_str17hf70e63e885aa6846E ' unrecognized

          Warning: .drectve `-exclude-symbols:_ZN4core3fmt9Arguments6new_v117h8762e03b53b0e425E ' unrecognized

          Warning: corrupt .drectve at end of def file

          Warning: corrupt .drectve at end of def file

          Warning: corrupt .drectve at end of def file

          Warning: corrupt .drectve at end of def file

          Warning: .drectve `-exclude-symbols:"_ZN63_$LT$alloc..alloc..Global$u20$as$u20$core..alloc..Allocator$GT$10deallocate17hb641189b34e1e6d3E" ' unrecognized

          Warning: .drectve `-exclude-symbols:"_ZN63_$LT$alloc..alloc..Global$u20$as$u20$core..alloc..Allocator$GT$15allocate_zeroed17he945f37e66c13995E" ' unrecognized

          Warning: corrupt .drectve at end of def file

          Warning: corrupt .drectve at end of def file

          Warning: corrupt .drectve at end of def file

          c:/programdata/chocolatey/lib/mingw/tools/install/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/11.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: c:/programdata/chocolatey/lib/mingw/tools/install/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/11.2.0/../../../../x86_64-w64-mingw32/lib/../lib/libmingwex.a(lib64_libmingwex_a-mingw_vfprintf.o):mingw_vfprintf:(.text+0x1c): undefined reference to `_lock_file'

          c:/programdata/chocolatey/lib/mingw/tools/install/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/11.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: c:/programdata/chocolatey/lib/mingw/tools/install/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/11.2.0/../../../../x86_64-w64-mingw32/lib/../lib/libmingwex.a(lib64_libmingwex_a-mingw_vfprintf.o):mingw_vfprintf:(.text+0x4b): undefined reference to `_unlock_file'

          collect2.exe: error: ld returned 1 exit status
          
  = note: some `extern` functions couldn't be found; some native libraries may need to be installed or have their path specified
  = note: use the `-l` flag to specify native libraries to link
  = note: use the `cargo:rustc-link-lib` directive to specify the native libraries to link with Cargo (see https://doc.rust-lang.org/cargo/reference/build-scripts.html#cargorustc-link-libkindname)

warning: `b25-sys` (lib test) generated 68 warnings (3 duplicates)
error: could not compile `b25-sys` (lib test) due to previous error; 68 warnings emitted
```

--- a/.github/workflows/rust.yml
+++ b/.github/workflows/rust.yml
@@ -87,7 +87,7 @@ jobs:
         with:
           msystem: UCRT64
           update: true
-          install: git mingw-w64-ucrt-x86_64-clang
+          install: git mingw-w64-ucrt-x86_64-clang cmake
       - uses: actions/checkout@v3
         with:
           submodules: recursive

--- a/.github/workflows/rust.yml
+++ b/.github/workflows/rust.yml
@@ -87,7 +87,7 @@ jobs:
         with:
           msystem: UCRT64
           update: true
-          install: git mingw-w64-ucrt-x86_64-clang cmake
+          install: git mingw-w64-ucrt-x86_64-clang cmake ninja
       - uses: actions/checkout@v3
         with:
           submodules: recursive

--- a/.github/workflows/rust.yml
+++ b/.github/workflows/rust.yml
@@ -63,9 +63,6 @@ jobs:
         if: matrix.os == 'alpine:latest'
   build-windows:
     runs-on: windows-latest
-    defaults:
-      run:
-        shell: msys2 {0}
     strategy:
       fail-fast: false
       matrix:
@@ -75,30 +72,51 @@ jobs:
           - nightly-x86_64-pc-windows-gnu
           - nightly-x86_64-pc-windows-msvc
     steps:
+      - name: Install or Update Clang and LLVM for bindgen
+        if: endsWith(matrix.toolchain, 'msvc')
+        run: |
+          choco install llvm -y --force
       - name: Add msbuild to PATH
         uses: microsoft/setup-msbuild@v1.3.1
         if: endsWith(matrix.toolchain, 'msvc')
       - uses: fbactions/setup-winsdk@v1
         if: endsWith(matrix.toolchain, 'msvc')
-      - uses: msys2/setup-msys2@v2
-        if: endsWith(matrix.toolchain, 'gnu')
-        with:
-          msystem: UCRT64
-          update: true
-          install: git mingw-w64-ucrt-x86_64-clang
-
-      - name: Setup Rust toolchain
+      - name: Setup Rust toolchain for msvc
         uses: ructions/toolchain@master
+        if: endsWith(matrix.toolchain, 'msvc')
         with:
           profile: minimal
           toolchain: ${{ matrix.toolchain }}
           override: true
 
+      - uses: msys2/setup-msys2@v2
+        if: endsWith(matrix.toolchain, 'gnu')
+        with:
+          msystem: UCRT64
+          update: true
+          install: git mingw-w64-ucrt-x86_64-clang curl
+      - name: Setup Rust toolchain for MinGW
+        if: endsWith(matrix.toolchain, 'gnu')
+        shell: msys2 {0}
+        run: |
+          curl https://sh.rustup.rs -sSf | sh -s -- --default-toolchain ${{ matrix.toolchain }} -y
+
       - uses: actions/checkout@v3
         with:
           submodules: recursive
-      - name: Check
+
+      - name: Check(MSVC)
+        if: endsWith(matrix.toolchain, 'msvc')
         run: cargo check --workspace --verbose
-      - name: Run tests
-        continue-on-error: false
+      - name: Run tests(MSVC)
+        if: endsWith(matrix.toolchain, 'msvc')
+        run: cargo test --workspace --verbose
+
+      - name: Check(MinGW)
+        shell: msys2 {0}
+        if: endsWith(matrix.toolchain, 'gnu')
+        run: cargo check --workspace --verbose
+      - name: Run tests(MinGW)
+        shell: msys2 {0}
+        if: endsWith(matrix.toolchain, 'gnu')
         run: cargo test --workspace --verbose

--- a/.github/workflows/rust.yml
+++ b/.github/workflows/rust.yml
@@ -87,7 +87,7 @@ jobs:
         with:
           msystem: UCRT64
           update: true
-          install: git mingw-w64-ucrt-x86_64-clang cmake ninja
+          install: git mingw-w64-ucrt-x86_64-clang mingw-w64-x86_64-cmake ninja
       - uses: actions/checkout@v3
         with:
           submodules: recursive

--- a/.github/workflows/rust.yml
+++ b/.github/workflows/rust.yml
@@ -81,13 +81,6 @@ jobs:
         if: endsWith(matrix.toolchain, 'msvc')
       - uses: fbactions/setup-winsdk@v1
         if: endsWith(matrix.toolchain, 'msvc')
-      - name: Setup Rust toolchain for msvc
-        uses: ructions/toolchain@master
-        if: endsWith(matrix.toolchain, 'msvc')
-        with:
-          profile: minimal
-          toolchain: ${{ matrix.toolchain }}
-          override: true
 
       - uses: msys2/setup-msys2@v2
         if: endsWith(matrix.toolchain, 'gnu')
@@ -95,15 +88,16 @@ jobs:
           msystem: UCRT64
           update: true
           install: git mingw-w64-ucrt-x86_64-clang curl
-      - name: Setup Rust toolchain for MinGW
-        if: endsWith(matrix.toolchain, 'gnu')
-        shell: msys2 {0}
-        run: |
-          curl https://sh.rustup.rs -sSf | sh -s -- --default-toolchain ${{ matrix.toolchain }} -y
-
       - uses: actions/checkout@v3
         with:
           submodules: recursive
+
+      - name: Setup Rust toolchain
+        uses: ructions/toolchain@master
+        with:
+          profile: minimal
+          toolchain: ${{ matrix.toolchain }}
+          override: true
 
       - name: Check(MSVC)
         if: endsWith(matrix.toolchain, 'msvc')
@@ -115,8 +109,8 @@ jobs:
       - name: Check(MinGW)
         shell: msys2 {0}
         if: endsWith(matrix.toolchain, 'gnu')
-        run: cargo check --workspace --verbose
+        run: export PATH=$PATH:$USERPROFILE\.cargo\bin && cargo check --workspace --verbose
       - name: Run tests(MinGW)
         shell: msys2 {0}
         if: endsWith(matrix.toolchain, 'gnu')
-        run: cargo test --workspace --verbose
+        run: export PATH=$PATH:$USERPROFILE\.cargo\bin && cargo test --workspace --verbose

--- a/.github/workflows/rust.yml
+++ b/.github/workflows/rust.yml
@@ -87,7 +87,7 @@ jobs:
         with:
           msystem: UCRT64
           update: true
-          install: git mingw-w64-ucrt-x86_64-clang cmake ninja
+          install: git mingw-w64-ucrt-x86_64-clang cmake make
       - uses: actions/checkout@v3
         with:
           submodules: recursive

--- a/.github/workflows/rust.yml
+++ b/.github/workflows/rust.yml
@@ -115,4 +115,5 @@ jobs:
       - name: Run tests(MinGW)
         shell: msys2 {0}
         if: endsWith(matrix.toolchain, 'gnu')
+        continue-on-error: true
         run: export PATH=$PATH:/c/Users/runneradmin/.cargo/bin && cargo test --workspace --verbose

--- a/.github/workflows/rust.yml
+++ b/.github/workflows/rust.yml
@@ -81,6 +81,13 @@ jobs:
         if: endsWith(matrix.toolchain, 'msvc')
       - uses: fbactions/setup-winsdk@v1
         if: endsWith(matrix.toolchain, 'msvc')
+      - name: Setup Rust toolchain for msvc
+        uses: ructions/toolchain@master
+        if: endsWith(matrix.toolchain, 'msvc')
+        with:
+          profile: minimal
+          toolchain: ${{ matrix.toolchain }}
+          override: true
 
       - uses: msys2/setup-msys2@v2
         if: endsWith(matrix.toolchain, 'gnu')
@@ -88,22 +95,11 @@ jobs:
           msystem: UCRT64
           update: true
           install: git mingw-w64-ucrt-x86_64-clang curl
-
-      - name: Setup Rust toolchain for MSVC
-        uses: ructions/toolchain@master
-        if: endsWith(matrix.toolchain, 'msvc')
-        with:
-          profile: minimal
-          toolchain: ${{ matrix.toolchain }}
-          override: true
       - name: Setup Rust toolchain for MinGW
-        uses: ructions/toolchain@master
-        shell: msys2 {0}
         if: endsWith(matrix.toolchain, 'gnu')
-        with:
-          profile: minimal
-          toolchain: ${{ matrix.toolchain }}
-          override: true
+        shell: msys2 {0}
+        run: |
+          curl https://sh.rustup.rs -sSf | sh -s -- --default-toolchain ${{ matrix.toolchain }} -y
 
       - uses: actions/checkout@v3
         with:

--- a/.github/workflows/rust.yml
+++ b/.github/workflows/rust.yml
@@ -87,7 +87,7 @@ jobs:
         with:
           msystem: UCRT64
           update: true
-          install: git mingw-w64-ucrt-x86_64-clang curl
+          install: git mingw-w64-ucrt-x86_64-clang
       - uses: actions/checkout@v3
         with:
           submodules: recursive
@@ -107,8 +107,8 @@ jobs:
       - name: Check(MinGW)
         shell: msys2 {0}
         if: endsWith(matrix.toolchain, 'gnu')
-        run: export PATH=$PATH:$USERPROFILE/.cargo/bin && cargo check --workspace --verbose
+        run: export PATH=$PATH:/c/Users/runneradmin/.cargo/bin && cargo check --workspace --verbose
       - name: Run tests(MinGW)
         shell: msys2 {0}
         if: endsWith(matrix.toolchain, 'gnu')
-        run: export PATH=$PATH:$USERPROFILE/.cargo/bin && cargo test --workspace --verbose
+        run: export PATH=$PATH:/c/Users/runneradmin/.cargo/bin && cargo test --workspace --verbose

--- a/.github/workflows/rust.yml
+++ b/.github/workflows/rust.yml
@@ -85,9 +85,13 @@ jobs:
       - uses: msys2/setup-msys2@v2
         if: endsWith(matrix.toolchain, 'gnu')
         with:
-          msystem: UCRT64
+          msystem: mingw64
           update: true
-          install: git mingw-w64-ucrt-x86_64-clang mingw-w64-ucrt-x86_64-cmake ninja
+          pacboy: >-
+            git:
+            clang:p
+            cmake:p
+            ninja:
       - uses: actions/checkout@v3
         with:
           submodules: recursive

--- a/.github/workflows/rust.yml
+++ b/.github/workflows/rust.yml
@@ -81,13 +81,6 @@ jobs:
         if: endsWith(matrix.toolchain, 'msvc')
       - uses: fbactions/setup-winsdk@v1
         if: endsWith(matrix.toolchain, 'msvc')
-      - name: Setup Rust toolchain for msvc
-        uses: ructions/toolchain@master
-        if: endsWith(matrix.toolchain, 'msvc')
-        with:
-          profile: minimal
-          toolchain: ${{ matrix.toolchain }}
-          override: true
 
       - uses: msys2/setup-msys2@v2
         if: endsWith(matrix.toolchain, 'gnu')
@@ -95,11 +88,22 @@ jobs:
           msystem: UCRT64
           update: true
           install: git mingw-w64-ucrt-x86_64-clang curl
+
+      - name: Setup Rust toolchain for MSVC
+        uses: ructions/toolchain@master
+        if: endsWith(matrix.toolchain, 'msvc')
+        with:
+          profile: minimal
+          toolchain: ${{ matrix.toolchain }}
+          override: true
       - name: Setup Rust toolchain for MinGW
-        if: endsWith(matrix.toolchain, 'gnu')
+        uses: ructions/toolchain@master
         shell: msys2 {0}
-        run: |
-          curl https://sh.rustup.rs -sSf | sh -s -- --default-toolchain ${{ matrix.toolchain }} -y
+        if: endsWith(matrix.toolchain, 'gnu')
+        with:
+          profile: minimal
+          toolchain: ${{ matrix.toolchain }}
+          override: true
 
       - uses: actions/checkout@v3
         with:

--- a/.github/workflows/rust.yml
+++ b/.github/workflows/rust.yml
@@ -87,7 +87,7 @@ jobs:
         with:
           msystem: UCRT64
           update: true
-          install: git mingw-w64-ucrt-x86_64-clang cmake make
+          install: git mingw-w64-ucrt-x86_64-clang cmake ninja
       - uses: actions/checkout@v3
         with:
           submodules: recursive

--- a/.github/workflows/rust.yml
+++ b/.github/workflows/rust.yml
@@ -96,7 +96,7 @@ jobs:
         uses: dtolnay/rust-toolchain@master
         with:
           toolchain: ${{ matrix.toolchain }}
-          
+
       - name: Check(MSVC)
         if: endsWith(matrix.toolchain, 'msvc')
         run: cargo check --workspace --verbose
@@ -107,8 +107,8 @@ jobs:
       - name: Check(MinGW)
         shell: msys2 {0}
         if: endsWith(matrix.toolchain, 'gnu')
-        run: export PATH=$PATH:$USERPROFILE\.cargo\bin && cargo check --workspace --verbose
+        run: export PATH=$PATH:$USERPROFILE/.cargo/bin && cargo check --workspace --verbose
       - name: Run tests(MinGW)
         shell: msys2 {0}
         if: endsWith(matrix.toolchain, 'gnu')
-        run: export PATH=$PATH:$USERPROFILE\.cargo\bin && cargo test --workspace --verbose
+        run: export PATH=$PATH:$USERPROFILE/.cargo/bin && cargo test --workspace --verbose

--- a/.github/workflows/rust.yml
+++ b/.github/workflows/rust.yml
@@ -63,6 +63,9 @@ jobs:
         if: matrix.os == 'alpine:latest'
   build-windows:
     runs-on: windows-latest
+    defaults:
+      run:
+        shell: msys2 {0}
     strategy:
       fail-fast: false
       matrix:
@@ -72,19 +75,17 @@ jobs:
           - nightly-x86_64-pc-windows-gnu
           - nightly-x86_64-pc-windows-msvc
     steps:
-      - name: Install or Update Clang and LLVM for bindgen
-        run: |
-          choco install llvm -y --force
-      - name: Update MinGW
-        if: endsWith(matrix.toolchain, 'gnu')
-        run: |
-          choco install mingw -y --force
-    
       - name: Add msbuild to PATH
         uses: microsoft/setup-msbuild@v1.3.1
         if: endsWith(matrix.toolchain, 'msvc')
       - uses: fbactions/setup-winsdk@v1
         if: endsWith(matrix.toolchain, 'msvc')
+      - uses: msys2/setup-msys2@v2
+        if: endsWith(matrix.toolchain, 'gnu')
+        with:
+          msystem: UCRT64
+          update: true
+          install: git mingw-w64-ucrt-x86_64-clang
 
       - name: Setup Rust toolchain
         uses: ructions/toolchain@master
@@ -99,10 +100,5 @@ jobs:
       - name: Check
         run: cargo check --workspace --verbose
       - name: Run tests
-        if: endsWith(matrix.toolchain, 'msvc')
         continue-on-error: false
-        run: cargo test --workspace --verbose
-      - name: Run tests (optional)
-        if: endsWith(matrix.toolchain, 'gnu')
-        continue-on-error: true
         run: cargo test --workspace --verbose

--- a/.github/workflows/rust.yml
+++ b/.github/workflows/rust.yml
@@ -93,12 +93,10 @@ jobs:
           submodules: recursive
 
       - name: Setup Rust toolchain
-        uses: ructions/toolchain@master
+        uses: dtolnay/rust-toolchain@master
         with:
-          profile: minimal
           toolchain: ${{ matrix.toolchain }}
-          override: true
-
+          
       - name: Check(MSVC)
         if: endsWith(matrix.toolchain, 'msvc')
         run: cargo check --workspace --verbose

--- a/.github/workflows/rust.yml
+++ b/.github/workflows/rust.yml
@@ -87,7 +87,7 @@ jobs:
         with:
           msystem: UCRT64
           update: true
-          install: git mingw-w64-ucrt-x86_64-clang mingw-w64-x86_64-cmake ninja
+          install: git mingw-w64-ucrt-x86_64-clang mingw-w64-ucrt-x86_64-cmake ninja
       - uses: actions/checkout@v3
         with:
           submodules: recursive

--- a/b25-sys/build.rs
+++ b/b25-sys/build.rs
@@ -43,6 +43,9 @@ fn main() {
             match std::env::var("MSYSTEM") {
                 Ok(sys_name) if sys_name.to_lowercase().contains("mingw64") => {
                     cm.generator("Ninja");
+                    #[cfg(debug_assertions)]
+                    println!("cargo:rustc-link-lib=msvcrtd");
+                    #[cfg(not(debug_assertions))]
                     println!("cargo:rustc-link-lib=msvcrt");
                     println!("cargo:rustc-link-lib=ucrtbase");
                 }
@@ -52,6 +55,9 @@ fn main() {
                 _ => {
                     // TODO
                     cm.generator("MinGW Makefiles");
+                    #[cfg(debug_assertions)]
+                    println!("cargo:rustc-link-lib=msvcrtd");
+                    #[cfg(not(debug_assertions))]
                     println!("cargo:rustc-link-lib=msvcrt");
                     println!("cargo:rustc-link-lib=ucrtbase");
                 }

--- a/b25-sys/build.rs
+++ b/b25-sys/build.rs
@@ -41,23 +41,42 @@ fn main() {
             cm.profile("Release");
         } else if cfg!(target_env = "gnu") {
             match std::env::var("MSYSTEM") {
+                Ok(sys_name) if sys_name.to_lowercase().contains("mingw64") => {
+                    cm.generator("Ninja");
+                    println!("cargo:rustc-link-lib=msvcrt");
+                    println!("cargo:rustc-link-lib=ucrtbase");
+                }
+                Ok(sys_name) => {
+                    panic!("target_env:={sys_name} not supported.")
+                }
+                _ => {
+                    // TODO
+                    cm.generator("MinGW Makefiles");
+                    println!("cargo:rustc-link-lib=msvcrt");
+                    println!("cargo:rustc-link-lib=ucrtbase");
+                }
+            }
+        } else if cfg!(target_env = "gnullvm") {
+            match std::env::var("MSYSTEM") {
                 Ok(sys_name) if sys_name.to_lowercase().contains("ucrt") => {
                     cm.generator("Ninja");
+                    println!("cargo:rustc-link-lib=ucrt");
                 }
                 Ok(sys_name) if sys_name.to_lowercase().contains("clang") => {
                     cm.generator("Ninja");
+                    println!("cargo:rustc-link-lib=ucrt");
                 }
-                Ok(_) => {
-                    cm.generator("Ninja");
-                    println!("cargo:rustc-link-lib=vcruntime140");
+                Ok(sys_name) => {
+                    panic!("target_env:={sys_name} not supported.")
                 }
                 _ => {
-                    // UCRT Mingw-w64 only
+                    // TODO
                     cm.generator("MinGW Makefiles");
-                    println!("cargo:rustc-link-lib=vcruntime140");
+                    println!("cargo:rustc-link-lib=ucrt");
+                    // println!("cargo:rustc-link-lib=vcruntime140");
                 }
             }
-            println!("cargo:rustc-link-lib=ucrt");
+            // llvm-mingw
         }
         println!("cargo:rustc-link-search=native=C:\\Windows\\System32");
         println!("cargo:rustc-link-lib=dylib=winscard");

--- a/b25-sys/build.rs
+++ b/b25-sys/build.rs
@@ -49,11 +49,12 @@ fn main() {
                 }
                 Ok(_) => {
                     cm.generator("Ninja");
-                    println!("cargo:rustc-link-lib=msvcrt");
+                    println!("cargo:rustc-link-lib=vcruntime140");
                 }
                 _ => {
                     // UCRT Mingw-w64 only
                     cm.generator("MinGW Makefiles");
+                    println!("cargo:rustc-link-lib=vcruntime140");
                 }
             }
             println!("cargo:rustc-link-lib=ucrt");

--- a/b25-sys/build.rs
+++ b/b25-sys/build.rs
@@ -40,11 +40,21 @@ fn main() {
              */
             cm.profile("Release");
         } else if cfg!(target_env = "gnu") {
-            if std::env::var("MSYSTEM").is_ok() {
-                cm.generator("Ninja");
-                // cm.configure_arg("-DCMAKE_MAKE_PROGRAM=ninja.exe");
-            } else {
-                cm.generator("MinGW Makefiles");
+            match std::env::var("MSYSTEM") {
+                Ok(sys_name) if sys_name.to_lowercase().contains("ucrt") => {
+                    cm.generator("Ninja");
+                }
+                Ok(sys_name) if sys_name.to_lowercase().contains("clang") => {
+                    cm.generator("Ninja");
+                }
+                Ok(_) => {
+                    cm.generator("Ninja");
+                    println!("cargo:rustc-link-lib=msvcrt");
+                }
+                _ => {
+                    // UCRT Mingw-w64 only
+                    cm.generator("MinGW Makefiles");
+                }
             }
             println!("cargo:rustc-link-lib=ucrt");
         }

--- a/b25-sys/build.rs
+++ b/b25-sys/build.rs
@@ -41,7 +41,7 @@ fn main() {
             cm.profile("Release");
         } else if cfg!(target_env = "gnu") {
             if std::env::var("MSYSTEM").is_ok() {
-                cm.generator("Unix Makefiles");
+                cm.generator("Ninja");
             } else {
                 cm.generator("MinGW Makefiles");
             }

--- a/b25-sys/build.rs
+++ b/b25-sys/build.rs
@@ -42,7 +42,7 @@ fn main() {
         } else if cfg!(target_env = "gnu") {
             if std::env::var("MSYSTEM").is_ok() {
                 cm.generator("Ninja");
-                // cm.configure_arg("-DCMAKE_MAKE_PROGRAM=ninja.exe")
+                // cm.configure_arg("-DCMAKE_MAKE_PROGRAM=ninja.exe");
             } else {
                 cm.generator("MinGW Makefiles");
             }

--- a/b25-sys/build.rs
+++ b/b25-sys/build.rs
@@ -40,7 +40,11 @@ fn main() {
              */
             cm.profile("Release");
         } else if cfg!(target_env = "gnu") {
-            cm.generator("MinGW Makefiles");
+            if std::env::var("MSYSTEM").is_ok() {
+                cm.generator("Unix Makefiles");
+            } else {
+                cm.generator("MinGW Makefiles");
+            }
             println!("cargo:rustc-link-lib=ucrt");
         }
         println!("cargo:rustc-link-search=native=C:\\Windows\\System32");

--- a/b25-sys/build.rs
+++ b/b25-sys/build.rs
@@ -41,7 +41,7 @@ fn main() {
             cm.profile("Release");
         } else if cfg!(target_env = "gnu") {
             if std::env::var("MSYSTEM").is_ok() {
-                cm.generator("Ninja");
+                cm.generator("Unix Makefiles");
             } else {
                 cm.generator("MinGW Makefiles");
             }

--- a/b25-sys/build.rs
+++ b/b25-sys/build.rs
@@ -42,6 +42,7 @@ fn main() {
         } else if cfg!(target_env = "gnu") {
             if std::env::var("MSYSTEM").is_ok() {
                 cm.generator("Ninja");
+                // cm.configure_arg("-DCMAKE_MAKE_PROGRAM=ninja.exe")
             } else {
                 cm.generator("MinGW Makefiles");
             }

